diff --git a/lix/libfetchers/git.cc b/lix/libfetchers/git.cc
index 36da76928..fd6fa2b5e 100644
--- a/lix/libfetchers/git.cc
+++ b/lix/libfetchers/git.cc
@@ -1000,7 +1000,7 @@ struct GitInputScheme : InputScheme
             });
             Finally const _wait([&] { proc.waitAndCheck(); });
 
-            TRY_AWAIT(unpackTarfile(_input.toURLString(), *proc.getStdout(), tmpDir));
+            TRY_AWAIT(unpackTarfile(*proc.getStdout(), tmpDir));
         }
 
         auto storePath = TRY_AWAIT(
diff --git a/lix/libfetchers/tarball.cc b/lix/libfetchers/tarball.cc
index 3ea47684d..d19dbebef 100644
--- a/lix/libfetchers/tarball.cc
+++ b/lix/libfetchers/tarball.cc
@@ -195,7 +195,7 @@ try {
     } else {
         Path tmpDir = createTempDir();
         AutoDelete autoDelete(tmpDir, true);
-        unpackTarfile(url, store->toRealPath(res.storePath), tmpDir);
+        unpackTarfile(store->toRealPath(res.storePath), tmpDir);
         auto members = readDirectory(tmpDir);
         if (members.size() != 1)
             throw nix::Error("tarball '%s' contains an unexpected number of top-level files", url);
diff --git a/lix/libstore/builtins/unpack-channel.cc b/lix/libstore/builtins/unpack-channel.cc
index a4f9addc7..4a351db83 100644
--- a/lix/libstore/builtins/unpack-channel.cc
+++ b/lix/libstore/builtins/unpack-channel.cc
@@ -7,7 +7,7 @@ void builtinUnpackChannel(const Path & out, const std::string & channelName, con
 {
     createDirs(out);
 
-    unpackTarfile(fmt("channel:%s", channelName), src, out);
+    unpackTarfile(src, out);
 
     auto entries = readDirectory(out);
     if (entries.size() != 1)
diff --git a/lix/libutil/compression.cc b/lix/libutil/compression.cc
index c47137613..b4f9b558f 100644
--- a/lix/libutil/compression.cc
+++ b/lix/libutil/compression.cc
@@ -61,8 +61,7 @@ struct ArchiveDecompressionSource : Source
     size_t read(char * data, size_t len) override {
         struct archive_entry * ae;
         if (!archive) {
-            // note: this name is not actually used for anything, and we have no way of guessing it
-            archive = std::make_unique<TarArchive>("", *src, true);
+            archive = std::make_unique<TarArchive>(*src, true);
             this->archive->check(
                 archive_read_next_header(this->archive->archive.get(), &ae),
                 "failed to read header (%s)"
diff --git a/lix/libutil/logging.hh b/lix/libutil/logging.hh
index 8ea3ec63f..9cb034563 100644
--- a/lix/libutil/logging.hh
+++ b/lix/libutil/logging.hh
@@ -14,19 +14,8 @@ namespace nix {
 
 enum ActivityType : uint8_t {
     actUnknown = 0,
-
-    /** Fields:
-     * 0: string: store path being copied
-     * 1: string: URI of source store
-     * 2: string: URI of destination store
-     */
     actCopyPath = 100,
-
-    /** Fields:
-     * 0: string: URI of the file being transfered
-     */
     actFileTransfer = 101,
-
     actRealise = 102,
     actCopyPaths = 103,
     actBuilds = 104,
@@ -38,7 +27,6 @@ enum ActivityType : uint8_t {
      * 3: int: nrRounds, not used anymore always 1?
      */
     actBuild = 105,
-
     actOptimiseStore = 106,
     actVerifyPaths = 107,
 
@@ -58,7 +46,6 @@ enum ActivityType : uint8_t {
      * 0: string: store path
      */
     actPostBuildHook = 110,
-
     actBuildWaiting = 111,
 };
 
diff --git a/lix/libutil/tarfile.cc b/lix/libutil/tarfile.cc
index 2c9245d3b..0ea05d950 100644
--- a/lix/libutil/tarfile.cc
+++ b/lix/libutil/tarfile.cc
@@ -4,7 +4,6 @@
 
 #include "async-io.hh"
 #include "file-descriptor.hh"
-#include "libutil/error.hh"
 #include "lix/libutil/c-calls.hh"
 #include "lix/libutil/charptr-cast.hh"
 #include "lix/libutil/file-system.hh"
@@ -50,11 +49,10 @@ void TarArchive::check(int err, const std::string & reason)
     }
 }
 
-TarArchive::TarArchive(std::string name, Source & source, bool raw)
+TarArchive::TarArchive(Source & source, bool raw)
     : archive{archive_read_new()}
     , source(&source)
     , buffer(65536)
-    , name(name)
 {
     if (!raw) {
         archive_read_support_filter_all(archive.get());
@@ -73,9 +71,7 @@ TarArchive::TarArchive(std::string name, Source & source, bool raw)
     );
 }
 
-TarArchive::TarArchive(const Path & path, std::optional<std::string> name)
-    : archive{archive_read_new()}
-    , name(name.value_or(path))
+TarArchive::TarArchive(const Path & path) : archive{archive_read_new()}
 {
     archive_read_support_filter_all(archive.get());
     archive_read_support_format_all(archive.get());
@@ -95,8 +91,6 @@ static void extract_archive(TarArchive & archive, const Path & destDir)
 {
     requireCString(destDir);
 
-    logger->startActivity(lvlTalkative, actUnknown, fmt("unpacking tarball '%s'", archive.name));
-
     int flags =
         ARCHIVE_EXTRACT_TIME | ARCHIVE_EXTRACT_SECURE_SYMLINKS | ARCHIVE_EXTRACT_SECURE_NODOTDOT;
 
@@ -140,7 +134,7 @@ static void extract_archive(TarArchive & archive, const Path & destDir)
     archive.close();
 }
 
-kj::Promise<Result<void>> unpackTarfile(std::string name, AsyncInputStream & source, const Path & destDir)
+kj::Promise<Result<void>> unpackTarfile(AsyncInputStream & source, const Path & destDir)
 try {
     Pipe pipe;
     pipe.create();
@@ -149,7 +143,7 @@ try {
         std::launch::async,
         [&](AutoCloseFD fd) {
             FdSource source(fd.get());
-            auto archive = TarArchive(name, source);
+            auto archive = TarArchive(source);
 
             createDirs(destDir);
             extract_archive(archive, destDir);
@@ -165,9 +159,9 @@ try {
     co_return result::current_exception();
 }
 
-void unpackTarfile(std::string name, const Path & tarFile, const Path & destDir)
+void unpackTarfile(const Path & tarFile, const Path & destDir)
 {
-    auto archive = TarArchive(tarFile, name);
+    auto archive = TarArchive(tarFile);
 
     createDirs(destDir);
     extract_archive(archive, destDir);
diff --git a/lix/libutil/tarfile.hh b/lix/libutil/tarfile.hh
index 7b6185397..292221ba4 100644
--- a/lix/libutil/tarfile.hh
+++ b/lix/libutil/tarfile.hh
@@ -14,13 +14,12 @@ struct TarArchive
     std::unique_ptr<struct archive, decltype([](auto * p) { archive_read_free(p); })> archive;
     Source * source;
     std::vector<unsigned char> buffer;
-    std::string name;
 
     void check(int err, const std::string & reason = "failed to extract archive (%s)");
 
-    TarArchive(std::string name, Source & source, bool raw = false);
+    TarArchive(Source & source, bool raw = false);
 
-    TarArchive(const Path & path, std::optional<std::string> name = {});
+    TarArchive(const Path & path);
 
     /// disable copy constructor
     TarArchive(const TarArchive &) = delete;
@@ -28,7 +27,8 @@ struct TarArchive
     void close();
 };
 
-kj::Promise<Result<void>> unpackTarfile(std::string name, AsyncInputStream & source, const Path & destDir);
+kj::Promise<Result<void>> unpackTarfile(AsyncInputStream & source, const Path & destDir);
+
+void unpackTarfile(const Path & tarFile, const Path & destDir);
 
-void unpackTarfile(std::string name, const Path & tarFile, const Path & destDir);
 }
diff --git a/lix/nix/prefetch.cc b/lix/nix/prefetch.cc
index 85ce02205..855254532 100644
--- a/lix/nix/prefetch.cc
+++ b/lix/nix/prefetch.cc
@@ -108,9 +108,10 @@ std::tuple<StorePath, Hash> prefetchFile(
 
         /* Optionally unpack the file. */
         if (unpack) {
+            auto act = logger->startActivity(lvlChatty, actUnknown, fmt("unpacking '%s'", url));
             Path unpacked = (Path) tmpDir + "/unpacked";
             createDirs(unpacked);
-            unpackTarfile(url, tmpFile, unpacked);
+            unpackTarfile(tmpFile, unpacked);
 
             /* If the archive unpacks to a single file/directory, then use
                that as the top-level. */
diff --git a/tests/unit/libutil/tarfile.cc b/tests/unit/libutil/tarfile.cc
index a3f66e531..eaefaeba6 100644
--- a/tests/unit/libutil/tarfile.cc
+++ b/tests/unit/libutil/tarfile.cc
@@ -177,7 +177,7 @@ FileChecker TarFixture::extract()
     AsyncIoRoot aio;
     auto stream = AsyncStringInputStream{sink.s};
 
-    aio.blockOn(unpackTarfile("test", stream, tmpDir));
+    aio.blockOn(unpackTarfile(stream, tmpDir));
 
     return FileChecker{tmpDir};
 }
