diff --git a/lix/libmain/progress-bar.cc b/lix/libmain/progress-bar.cc
index 490b1105f..4deeb5f21 100644
--- a/lix/libmain/progress-bar.cc
+++ b/lix/libmain/progress-bar.cc
@@ -519,13 +519,21 @@ std::string ProgressBar::getStatus(State & state)
 
     showActivity(actBuilds, "%s built");
 
-    auto s1 = renderActivity(actCopyPaths, "%s copied");
-    auto s2 = renderActivity(actCopyPath, "%s MiB", "%.1f", MiB);
+    auto copiedPathsAmount = renderActivity(actCopyPaths, "%s copied");
+    auto copiedSize = renderActivity(actCopyPath, "%s MiB", "%.1f", MiB);
 
-    if (!s1.empty() || !s2.empty()) {
+    if (!copiedPathsAmount.empty() || !copiedSize.empty()) {
         if (!res.empty()) res += ", ";
-        if (s1.empty()) res += "0 copied"; else res += s1;
-        if (!s2.empty()) { res += " ("; res += s2; res += ')'; }
+        if (copiedPathsAmount.empty()) {
+            // note: we know copiedSize isn't empty cause otherwise we
+            //       wouldn't be in this outer if anyway
+            res += copiedSize + " copied";
+        } else {
+            res += copiedPathsAmount;
+            if (!copiedSize.empty()) {
+                res += " (" + copiedSize + ")";
+            }
+        }
     }
 
     showActivity(actFileTransfer, "%s MiB DL", "%.1f", MiB);
diff --git a/lix/libutil/compression.cc b/lix/libutil/compression.cc
index c47137613..b3093e2da 100644
--- a/lix/libutil/compression.cc
+++ b/lix/libutil/compression.cc
@@ -61,7 +61,6 @@ struct ArchiveDecompressionSource : Source
     size_t read(char * data, size_t len) override {
         struct archive_entry * ae;
         if (!archive) {
-            // note: this name is not actually used for anything, and we have no way of guessing it
             archive = std::make_unique<TarArchive>("", *src, true);
             this->archive->check(
                 archive_read_next_header(this->archive->archive.get(), &ae),
diff --git a/lix/libutil/logging.hh b/lix/libutil/logging.hh
index 8ea3ec63f..9cb034563 100644
--- a/lix/libutil/logging.hh
+++ b/lix/libutil/logging.hh
@@ -14,19 +14,8 @@ namespace nix {
 
 enum ActivityType : uint8_t {
     actUnknown = 0,
-
-    /** Fields:
-     * 0: string: store path being copied
-     * 1: string: URI of source store
-     * 2: string: URI of destination store
-     */
     actCopyPath = 100,
-
-    /** Fields:
-     * 0: string: URI of the file being transfered
-     */
     actFileTransfer = 101,
-
     actRealise = 102,
     actCopyPaths = 103,
     actBuilds = 104,
@@ -38,7 +27,6 @@ enum ActivityType : uint8_t {
      * 3: int: nrRounds, not used anymore always 1?
      */
     actBuild = 105,
-
     actOptimiseStore = 106,
     actVerifyPaths = 107,
 
@@ -58,7 +46,6 @@ enum ActivityType : uint8_t {
      * 0: string: store path
      */
     actPostBuildHook = 110,
-
     actBuildWaiting = 111,
 };
 
diff --git a/lix/libutil/tarfile.cc b/lix/libutil/tarfile.cc
index 2c9245d3b..da85edea6 100644
--- a/lix/libutil/tarfile.cc
+++ b/lix/libutil/tarfile.cc
@@ -95,7 +95,24 @@ static void extract_archive(TarArchive & archive, const Path & destDir)
 {
     requireCString(destDir);
 
-    logger->startActivity(lvlTalkative, actUnknown, fmt("unpacking tarball '%s'", archive.name));
+    auto unpackedSize =
+        logger->startActivity(
+            lvlInfo,
+            actCopyPath,
+            fmt("unpacking tarball '%s'", archive.name),
+            {
+                fmt("/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-%s", baseNameOf(archive.name)),
+                archive.name,
+                fmt("file://%s", destDir),
+            }
+        );
+
+    size_t processedBytes = 0;
+
+    // keeps track of the value of `processedBytes` the last time we reported progress
+    size_t lastLoggedBytes = 0;
+    // to avoid log spam, we only log at constant size increments (here, 100 KiB)
+    constexpr auto logIncrement = 100l * 1024;
 
     int flags =
         ARCHIVE_EXTRACT_TIME | ARCHIVE_EXTRACT_SECURE_SYMLINKS | ARCHIVE_EXTRACT_SECURE_NODOTDOT;
@@ -135,8 +152,22 @@ static void extract_archive(TarArchive & archive, const Path & destDir)
         }
 
         archive.check(archive_read_extract(archive.archive.get(), entry, flags));
+
+        // only count regular files (e.g. directories would double-count the size of the files under them)
+        if (archive_entry_filetype(entry) == AE_IFREG) {
+            processedBytes += archive_entry_size(entry);
+            // avoid log chatter by only progressing every once in a while
+            if ((processedBytes - lastLoggedBytes) > logIncrement) {
+                lastLoggedBytes = processedBytes;
+                // note: we discard the BufferState because these logs are not essential and
+                //       don't need to wait for flushing
+                (void) unpackedSize.progress(processedBytes);
+            }
+        }
     }
 
+    (void) unpackedSize.progress(processedBytes);
+
     archive.close();
 }
 
