commit 5958328964f150ca9e68c49d3be44529ec1831e0
Author: blokyk <courvoisier.zoe@gmail.com>
Date:   Mon May 25 02:36:09 2026 +0200

    on-primary-monitor

diff --git a/egui/Cargo.toml b/egui/Cargo.toml
index 98f4ced..fa00ad6 100644
--- a/egui/Cargo.toml
+++ b/egui/Cargo.toml
@@ -12,6 +12,7 @@ license.workspace = true
 [dependencies]
 arrayvec = "0.7.6"
 eframe = { version = "0.34.2", default-features = false, features = ["glow"] }
+display-info = "0.5.9"
 font-kit = { version = "0.14.3", optional = true }
 image = "0.25.10"
 itoa = "1.0.18"
diff --git a/egui/src/main.rs b/egui/src/main.rs
index 968c93e..b6f70e3 100644
--- a/egui/src/main.rs
+++ b/egui/src/main.rs
@@ -23,11 +23,12 @@ use eframe::{
     egui,
     egui::{
         CentralPanel, Event, FontId, Frame, Image, Key, Label, Margin, Modifiers, Panel, Popup,
-        PopupCloseBehavior, Pos2, Response, RichText, ScrollArea, Sense, Stroke, TextEdit,
+        PopupCloseBehavior, Response, RichText, ScrollArea, Sense, Stroke, TextEdit,
         TextFormat, ThemePreference, Ui, Vec2, ViewportBuilder, ViewportCommand, Widget,
         text::{LayoutJob, LayoutSection},
     },
 };
+use display_info::DisplayInfo;
 use image::ImageReader;
 use itoa::Integer;
 use ringboard_sdk::{
@@ -50,6 +51,7 @@ use crate::{
     startup::{maintain_single_instance, maybe_open_existing_instance_and_exit, sleep_file_name},
 };
 
+
 mod startup;
 
 #[cfg(feature = "trace")]
@@ -68,6 +70,9 @@ fn main() -> Result<(), eframe::Error> {
         None
     };
 
+    let displays = DisplayInfo::all().unwrap();
+    let primary = displays.iter().find(|d| d.is_primary).unwrap();
+
     let stop = Arc::new(AtomicBool::new(false));
     let result = eframe::run_native(
         concat!("Ringboard v", env!("CARGO_PKG_VERSION")),
@@ -76,7 +81,7 @@ fn main() -> Result<(), eframe::Error> {
                 .with_app_id("ringboard-egui")
                 .with_min_inner_size(Vec2::splat(100.))
                 .with_inner_size(Vec2::new(666., 777.))
-                .with_position(Pos2::ZERO),
+                .with_position([primary.x as f32, primary.y as f32]),
             ..Default::default()
         },
         Box::new(|cc| {
